lib/dslash_coarse.in.cpp

```diff
@@ -1,45 +1,122 @@
+    } else {
+      if constexpr (sizeof...(N) > 0) {
+        ApplyCoarseMma<dagger, Nc>(out, inA, inB, Y, X, kappa, parity, dslash, clover, commDim, halo_precision,
+                                   IntList<N...>());
+      } else {
+        errorQuda("nVec = %d has not been instantiated", out[0].Nvec());
+      }
+    }
+  }
+
+  template <bool use_mma, int Nc, int... N>
   void ApplyCoarse(cvector_ref<ColorSpinorField> &out, cvector_ref<const ColorSpinorField> &inA,
                    cvector_ref<const ColorSpinorField> &inB, const GaugeField &Y, const GaugeField &X, double kappa,
                    int parity, bool dslash, bool clover, bool dagger, const int *commDim, QudaPrecision halo_precision,
                    IntList<Nc, N...>)
   {
-    if (inA[0].Ncolor() == Nc) {
+    int nColor = use_mma ? inA[0].Ncolor() / inA[0].Nvec() : inA[0].Ncolor();
+    if (nColor == Nc) {
       if (dagger)
-        ApplyCoarse<true, Nc>(out, inA, inB, Y, X, kappa, parity, dslash, clover, commDim, halo_precision);
+        if constexpr (use_mma)
+          ApplyCoarseMma<true, Nc>(out, inA, inB, Y, X, kappa, parity, dslash, clover, commDim, halo_precision,
+                                   IntList<@QUDA_MULTIGRID_MRHS_LIST@>());
+        else
+          ApplyCoarse<true, Nc>(out, inA, inB, Y, X, kappa, parity, dslash, clover, commDim, halo_precision);
+      else if constexpr (use_mma)
+        ApplyCoarseMma<false, Nc>(out, inA, inB, Y, X, kappa, parity, dslash, clover, commDim, halo_precision,
+                                  IntList<@QUDA_MULTIGRID_MRHS_LIST@>());
       else
         ApplyCoarse<false, Nc>(out, inA, inB, Y, X, kappa, parity, dslash, clover, commDim, halo_precision);
     } else {
       if constexpr (sizeof...(N) > 0) {
-        ApplyCoarse(out, inA, inB, Y, X, kappa, parity, dslash, clover, dagger, commDim, halo_precision, IntList<N...>());
+        ApplyCoarse<use_mma>(out, inA, inB, Y, X, kappa, parity, dslash, clover, dagger, commDim, halo_precision,
+                             IntList<N...>());
       } else {
         errorQuda("Nc = %d has not been instantiated", inA[0].Ncolor());
       }
     }
   }
 
+  auto create_gauge_copy(const GaugeField &X, QudaGaugeFieldOrder order, bool copy_content)
+  {
+    GaugeField *output;
+    if (X.Order() == order) {
+      output = const_cast<GaugeField *>(&X);
+    } else {
+      GaugeFieldParam param(X);
+      param.order = order;
+      param.location = QUDA_CUDA_FIELD_LOCATION;
+      output = static_cast<GaugeField *>(cudaGaugeField::Create(param));
+      if (copy_content) { output->copy(X); }
+    }
+    return output;
+  }
+
+  template <class F> auto create_color_spinor_copy(cvector_ref<F> &fs, QudaFieldOrder order)
+  {
+    ColorSpinorParam param(fs[0]);
+    int nVec = (fs.size() + 7) / 8 * 8;   // Make a multiple of 8
+    param.nColor = fs[0].Ncolor() * nVec; // Ask Kate why we need * in.size() here
```

Comment:
Clean up the comment

Reply:
https://github.com/lattice/quda/pull/1355/commits/ef5c9921a50a31bb86a0ed08739cb17ea6a85cca